File: day4/day4.py
import re

rules = {
    'byr': lambda s: s.isdigit() and 1920 <= int(s) <= 2002,
    'iyr': lambda s: s.isdigit() and 2010 <= int(s) <= 2020,
    'eyr': lambda s: s.isdigit() and 2020 <= int(s) <= 2030,
    'hgt': lambda s: s[:-2].isdigit() and (
        (s.endswith('cm') and 150 <= int(s[:-2]) <= 193) or
        (s.endswith('in') and 59 <= int(s[:-2]) <= 76)),
    'hcl': lambda s: re.match(r'#[0-9a-f]{6}', s),
    'ecl': lambda s: re.match(r'amb|blu|brn|gry|grn|hzl|oth', s),
    'pid': lambda s: s.isdigit() and len(s) == 9,
}

data = [dict(re.findall(r'([a-z]{3}):(\S+)', block))
        for block in open('day4.txt').read().split('\n\n')]
print(sum(all(k in item for k in rules) for item in data))
print(sum(all(v(item.get(k, '')) for k, v in rules.items()) for item in data))

File: 2020/day4/day4.kt
import java.io.File

val rules = mapOf(
    "byr" to { s: String -> s.toIntOrNull() in 1920..2002 },
    "iyr" to { s: String -> s.toIntOrNull() in 2010..2020 },
    "eyr" to { s: String -> s.toIntOrNull() in 2020..2030 },
    "hgt" to { s: String -> when {
        s.endsWith("cm") -> s.dropLast(2).toIntOrNull() in 150..193
        s.endsWith("in") -> s.dropLast(2).toIntOrNull() in 59..76
        else -> false
    }},
    "hcl" to { s: String -> Regex("#[0-9a-f]{6}").matches(s) },
    "ecl" to { s: String -> Regex("amb|blu|brn|gry|grn|hzl|oth").matches(s) },
    "pid" to { s: String -> s.length == 9 && s.toIntOrNull() != null },
)

fun main() {
    val input = File("day4.txt").readText().split("\n\n")
        .map { Regex("""(\w+):(\S+)""").findAll(it)
        .map { it.groupValues[1] to it.groupValues[2] }.toMap() }
    println(input.count { item -> rules.keys.all { item.contains(it) } })
    println(input.count { item -> rules.entries.all {
        (key, test) -> test(item.getOrDefault(key, "")) } })
}
